extensions/stats/plugin.h

```diff
@@ -74,70 +74,55 @@ using google::protobuf::util::Status;
-  DIMENSION(destination_workload_namespace) \
-  DIMENSION(destination_principal)          \
-  DIMENSION(destination_app)                \
-  DIMENSION(destination_version)            \
-  DIMENSION(destination_service_host)       \
-  DIMENSION(destination_service_name)       \
-  DIMENSION(destination_service_namespace)  \
-  DIMENSION(request_protocol)               \
-  DIMENSION(response_code)                  \
-  DIMENSION(response_flags)                 \
-  DIMENSION(connection_security_policy)
+#define STD_ISTIO_DIMENSIONS(FIELD_FUNC)     \
+  FIELD_FUNC(reporter)                       \
+  FIELD_FUNC(source_workload)                \
+  FIELD_FUNC(source_workload_namespace)      \
+  FIELD_FUNC(source_principal)               \
+  FIELD_FUNC(source_app)                     \
+  FIELD_FUNC(source_version)                 \
+  FIELD_FUNC(destination_workload)           \
+  FIELD_FUNC(destination_workload_namespace) \
+  FIELD_FUNC(destination_principal)          \
+  FIELD_FUNC(destination_app)                \
+  FIELD_FUNC(destination_version)            \
+  FIELD_FUNC(destination_service_host)       \
+  FIELD_FUNC(destination_service_name)       \
+  FIELD_FUNC(destination_service_namespace)  \
+  FIELD_FUNC(request_protocol)               \
+  FIELD_FUNC(response_code)                  \
+  FIELD_FUNC(response_flags)                 \
+  FIELD_FUNC(connection_security_policy)
 
 struct IstioDimensions {
-#define DIMENSION(name) std::string(name);
-  ISTIO_DIMENSIONS
-#undef DIMENSION
+#define DEFINE_FIELD(name) std::string(name);
+
+  STD_ISTIO_DIMENSIONS(DEFINE_FIELD)
 
   // utility fields
   std::vector<std::string> vals;
   bool mapped = false;
   bool outbound = false;
 
-  // ordered dimension list is used by the metrics API.
-  static const std::vector<std::string> list() {
-    return std::vector<std::string>{
-#define DIMENSION(name) #name,
-        ISTIO_DIMENSIONS
-#undef DIMENSION
-    };
-  }
-
   // Ordered dimension list is used by the metrics API.
   static std::vector<MetricTag> metricTags() {
-    return std::vector<MetricTag>{
-#define DIMENSION(name) {#name, MetricTag::TagType::String},
-        ISTIO_DIMENSIONS
-#undef DIMENSION
-    };
+#define DEFINE_METRIC(name) {#name, MetricTag::TagType::String},
+    return std::vector<MetricTag>{STD_ISTIO_DIMENSIONS(DEFINE_METRIC)};
   }
 
   // values is used on the datapath, only when new dimensions are found.
   std::vector<std::string> values() {
-    return std::vector<std::string>{
-#define DIMENSION(name) absl::StrReplaceAll(name, HACK_VALUES_REPLACEMENTS),
-        ISTIO_DIMENSIONS
-#undef DIMENSION
-    };
+#define REPLACE_VALUES(name) \
+  absl::StrReplaceAll(name, HACK_VALUES_REPLACEMENTS),
+    return std::vector<std::string>{STD_ISTIO_DIMENSIONS(REPLACE_VALUES)};
   }
 
   void setFieldsUnknownIfEmpty() {
-#define DIMENSION(name) \
-  if ((name).empty()) { \
-    (name) = unknown;   \
+#define SET_IF_EMPTY(name) \
```

Comment:
Should this be `if not unknown`? For example if for one request a field is known and filled, the in following request the field becomes unknown, its value won't be overrode by `unknown` right?

Reply:
I think you might need an extra data structure to record which field is filled if dimensions are shared across requests?